router-tests/timeout_test.go

```diff
@@ -0,0 +1,132 @@
+package integration_test
+
+import (
+	"github.com/stretchr/testify/require"
+	"github.com/wundergraph/cosmo/router-tests/testenv"
+	"github.com/wundergraph/cosmo/router/core"
+	"github.com/wundergraph/cosmo/router/pkg/config"
+	"net/http"
+	"testing"
+	"time"
+)
+
+func TestTimeouts(t *testing.T) {
+	t.Parallel()
+
+	const queryEmployeeWithHobby = `{
+	  employee(id: 1) {
+		id
+		hobbies {
+		  ... on Gaming {
+			name
+		  }
+		}
+	  }
+	}`
+
+	const queryEmployeeWithNoHobby = `{
+	  employee(id: 1) {
+		id
+	  }
+	}`
+
+	t.Run("applies RequestTimeout", func(t *testing.T) {
+		t.Parallel()
+
+		hobbySubgraphSleep := testenv.SubgraphsConfig{
+			Hobbies: testenv.SubgraphConfig{
+				Middleware: func(handler http.Handler) http.Handler {
+					return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+						w.WriteHeader(http.StatusOK)
+						time.Sleep(5 * time.Millisecond) // Slow response
+						w.Write([]byte("Hello, world!"))
+					})
+				},
+			},
+		}
+
+		trafficConfig := config.TrafficShapingRules{
+			All: config.GlobalSubgraphRequestRule{
+				RequestTimeout: 10 * time.Millisecond,
+			},
+			Subgraphs: map[string]*config.GlobalSubgraphRequestRule{
+				"hobbies": {
+					RequestTimeout: 3 * time.Millisecond,
+				},
+			},
+		}
+		t.Run("applied subgraph timeout to request", func(t *testing.T) {
+			t.Parallel()
+
+			testenv.Run(t, &testenv.Config{
+				Subgraphs: hobbySubgraphSleep,
+				RouterOptions: []core.Option{
+					core.WithSubgraphTransportOptions(
+						core.NewSubgraphTransportOptions(trafficConfig)),
+				},
+			}, func(t *testing.T, xEnv *testenv.Environment) {
+				res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+					Query: queryEmployeeWithHobby,
+				})
+				require.Equal(t, `{"errors":[{"message":"Failed to fetch from Subgraph 'hobbies' at Path 'employee'."}],"data":{"employee":{"id":1,"hobbies":null}}}`, res.Body)
+			})
+		})
+
+		t.Run("Slow response exceeding timeout", func(t *testing.T) {
+			t.Parallel()
+
+			testenv.Run(t, &testenv.Config{
+				Subgraphs: hobbySubgraphSleep,
+				RouterOptions: []core.Option{
+					core.WithSubgraphTransportOptions(
+						core.NewSubgraphTransportOptions(trafficConfig)),
+				},
+			}, func(t *testing.T, xEnv *testenv.Environment) {
+				res := xEnv.MakeGraphQLRequestOK(testenv.GraphQLRequest{
+					Query: queryEmployeeWithNoHobby,
+				})
+				require.Equal(t, `{"data":{"employee":{"id":1}}}`, res.Body)
+			})
+		})
+	})
+
+	t.Run("ResponseHeaderTimeout exceeded", func(t *testing.T) {
```

Comment:
I only tested RequestTimeout and ResponseHeaderTimeout with integration tests. The other timeout options are more complicated to test separately by themselves, from what I had seen